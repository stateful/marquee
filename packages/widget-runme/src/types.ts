import type { ContextProperties } from '@vscode-marquee/utils'

export interface State {
  isInstalled: boolean
  notebooks: null | string[]
}
export interface Configuration {}

export interface Context extends ContextProperties<Configuration & State> {}
