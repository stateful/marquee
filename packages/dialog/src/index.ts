import DialogContainer from './DialogContainer'
import DialogTitle from './DialogTitle'

export { DialogContainer, DialogTitle }
