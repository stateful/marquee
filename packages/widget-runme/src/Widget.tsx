import React, { useContext }  from 'react'
import { Grid, Typography } from '@mui/material'
import wrapper, { Dragger, HeaderWrapper, NavIconDropdown } from '@vscode-marquee/widget'
import type { MarqueeWidgetProps } from '@vscode-marquee/widget'
import PopupState from 'material-ui-popup-state'

import InstallRunme from './components/InstallRunme'
import NotebookList from './components/NotebookList'
import RunmeContext, { RunmeProvider } from './Context'

const Runme = ({ ToggleFullScreen, fullscreenMode, minimizeNavIcon }: MarqueeWidgetProps) => {
  const { isInstalled } = useContext(RunmeContext)

  const NavButtons = () => (
    <Grid item>
      <Grid
        container
        justifyContent="right"
        direction={minimizeNavIcon ? 'column-reverse' : 'row'}
        spacing={1}
        alignItems="center"
        padding={minimizeNavIcon ? 0.5 : 0}
      >
        <Grid item>
          <ToggleFullScreen />
        </Grid>
        {!fullscreenMode &&
          <Grid item>
            <Dragger />
          </Grid>
        }
      </Grid>
    </Grid>
  )

  return (
    <>
      <HeaderWrapper>
        <Grid item>
          <Typography variant="subtitle1">Runme Markdowns</Typography>
        </Grid>
        {minimizeNavIcon ?
          <PopupState variant='popper' popupId='widget-runme'>
            {(popupState) => {
              return (
                <NavIconDropdown popupState={popupState}>
                  <NavButtons />
                </NavIconDropdown>
              )}}
          </PopupState>
          :
          <Grid item xs={8}>
            <NavButtons />
          </Grid>
        }
      </HeaderWrapper>
      <Grid item xs>
        <Grid
          container
          wrap="nowrap"
          direction="column"
          style={{ height: '100%' }}
        >
          <Grid item xs style={{ overflow: 'auto' }}>
            {isInstalled
              ? <NotebookList />
              : <InstallRunme />
            }
          </Grid>
        </Grid>
      </Grid>
    </>
  )
}

export default wrapper((props: any) => (
  <RunmeProvider>
    <Runme {...props} />
  </RunmeProvider>
), 'dependencies')
