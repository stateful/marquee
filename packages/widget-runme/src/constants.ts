import vscode from 'vscode'
import type { Configuration, State } from './types'

export const DEFAULT_CONFIGURATION: Configuration = {}
export const DEFAULT_STATE: State = {
  isInstalled: Boolean(vscode.extensions.all.find((ext) => ext.id === 'stateful.runme')),
  notebooks: null
}
