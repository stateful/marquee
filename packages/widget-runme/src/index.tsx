import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faPlay } from '@fortawesome/free-solid-svg-icons/faPlay'

import RunmeWidget from './Widget'

export default {
  name: 'runme',
  icon: <FontAwesomeIcon icon={faPlay} />,
  label: 'Runme',
  tags: ['workspace', 'projects', 'productivity'],
  description: 'Use dev-native markdown to craft interactive runbooks for VS Code.',
  component: RunmeWidget
}
