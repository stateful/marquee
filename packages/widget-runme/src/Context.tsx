import React from 'react'

import { connect, getEventListener, MarqueeWindow } from '@vscode-marquee/utils'
import { createContext } from 'react'
import { Configuration, Context, State } from './types'

declare const window: MarqueeWindow

const RunmeContext = createContext<Context>({} as any)
const WIDGET_ID = '@vscode-marquee/runme-widget'

interface Props {
  children?: React.ReactNode;
}

const RunmeProvider = ({ children }: Props) => {
  const widgetState = getEventListener<State & Configuration>(WIDGET_ID)
  const providerValues = connect<Configuration & State>(
    {
      ...window.marqueeStateConfiguration[WIDGET_ID].configuration,
      ...window.marqueeStateConfiguration[WIDGET_ID].state
    }, widgetState
  )

  return (
    <RunmeContext.Provider
      value={{ ...providerValues }}
    >
      { children }
    </RunmeContext.Provider>
  )
}

export default RunmeContext
export { RunmeProvider }
