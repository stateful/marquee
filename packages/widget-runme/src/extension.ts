import url from 'url'
import vscode from 'vscode'
import ExtensionManager, { Logger }  from '@vscode-marquee/utils/extension'

import { getExcludePattern } from './utils'
import { DEFAULT_CONFIGURATION, DEFAULT_STATE } from './constants'
import type { Configuration, State } from './types'

const STATE_KEY = 'widgets.runme'

export class RunmeExtensionManager extends ExtensionManager<State, Configuration> {
  #notebooks: string[] = []

  constructor (context: vscode.ExtensionContext) {
    super(
      context,
      STATE_KEY,
      DEFAULT_CONFIGURATION,
      DEFAULT_STATE
    )
    this._state = DEFAULT_STATE
    this.#init()

    const watcher = vscode.workspace.createFileSystemWatcher(
      '**/*.md',
      false,
      true,
      false
    )
    watcher.onDidCreate((uri) => this.#onFileChange(uri))
    watcher.onDidDelete((uri) => this.#onFileChange(uri, true))
  }

  async #init () {
    const aws = this.getActiveWorkspace()
    const excludePatterns = await getExcludePattern(aws)

    this.#notebooks = [
      ...await vscode.workspace.findFiles('**/*.md', `{${excludePatterns}}`),
      // ...await vscode.workspace.findFiles('**/*.mdx'),
    ].map((notebook) => url.pathToFileURL(notebook.fsPath).toString())
    this.updateState('notebooks', this.#notebooks, true)
  }

  #onFileChange (uri: vscode.Uri, wasDeleted?: boolean) {
    const notebookPath = uri.toString()
    const notebooks = JSON.parse(JSON.stringify(this.#notebooks))

    if (!wasDeleted) {
      notebooks.push(notebookPath)
    } else if (notebooks.includes(notebookPath)) {
      notebooks.splice(notebooks.indexOf(notebookPath), 1)
    } else {
      return
    }

    this.updateState('notebooks', notebooks, true)
    this.broadcast({ notebooks })
    Logger.info(wasDeleted
      ? `Runme notebook ("${notebookPath}") was deleted from file system`
      : `New notebook found in file system "${notebookPath}"`)
  }
}

export function activate (context: vscode.ExtensionContext) {
  const stateManager = new RunmeExtensionManager(context)
  return {
    marquee: {
      disposable: stateManager,
      defaultState: stateManager.state,
      defaultConfiguration: stateManager.configuration,
      setup: stateManager.setBroadcaster.bind(stateManager)
    }
  }
}
