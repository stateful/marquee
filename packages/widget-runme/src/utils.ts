import path from 'path'
import vscode from 'vscode'
import { Logger }  from '@vscode-marquee/utils/extension'
import type { Workspace } from '@vscode-marquee/utils'

export function mapGitIgnoreToGlobFolders (gitignoreContents: string[]): Array<string | undefined> {
  const entries = gitignoreContents
    .filter((entry: string) => entry)
    .map((entry: string) => entry.replace(/\s/g, ''))
    .map((entry: string) => {
      if (entry) {
        let firstChar = entry.charAt(0)
        if (firstChar === '!' || firstChar === '/') {
          entry = entry.substring(1, entry.length)
          firstChar = entry.charAt(0)
        }
        const hasExtension = path.extname(entry)
        const slashPlacement = entry.charAt(entry.length - 1)
        if (slashPlacement === '/') {
          return `**/${entry}**`
        }
        if (hasExtension || ['.', '*', '#'].includes(firstChar)) {
          return
        }
        return `**/${entry}/**`
      }
    }).filter((entry: string | undefined) => entry)

  return [...new Set(entries)]
}

export async function getExcludePattern (aws: Workspace | null) {
  let excludePatterns = ''

  if (!aws) {
    return excludePatterns
  }

  const gitIgnoreUri = vscode.Uri.joinPath(vscode.Uri.parse(aws.path), '.gitignore')
  try {
    const gitignoreStat = await vscode.workspace.fs.stat(gitIgnoreUri)
    if (gitignoreStat.type === vscode.FileType.File) {
      const ignoreList = await vscode.workspace.openTextDocument(gitIgnoreUri)
      const patterns = mapGitIgnoreToGlobFolders(ignoreList.getText().split('\n'))
      excludePatterns = patterns.join(',')
    }
  } catch (err: unknown) {
    Logger.warn(`Failed to detect or load .gitignore file: ${(err as Error).message}`)
  }

  return excludePatterns
}
