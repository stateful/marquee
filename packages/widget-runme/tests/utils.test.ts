import vscode from 'vscode'
import { mapGitIgnoreToGlobFolders, getExcludePattern } from '../src/utils'

const gitignore = `dist
out
node_modules
.vscode-test/
*.vsix
.DS_Store
*.log
packages/**/build
website/public/*.js
website/public/*.js.map`

test('mapGitIgnoreToGlobFolders', () => {
  expect(mapGitIgnoreToGlobFolders([
    'dist',
    'out',
    'node_modules',
    '.vscode-test/',
    '*.vsix',
    '.DS_Store',
    '*.log',
    'packages/**/build',
    'website/public/*.js',
    'website/public/*.js.map'
  ])).toMatchSnapshot()
})

test('getExcludePattern', async () => {
  expect(await getExcludePattern(null)).toBe('')
  expect(await getExcludePattern({ path: '/foo/bar' } as any)).toEqual('')
  ;(vscode.workspace.fs.stat as jest.Mock).mockResolvedValue({ type: 3 })
  ;(vscode.workspace.openTextDocument as jest.Mock).mockResolvedValue({
    getText: jest.fn().mockReturnValue(gitignore)
  })
  expect(await getExcludePattern({ path: '/foo/bar' } as any)).toEqual('')
  ;(vscode.workspace.fs.stat as jest.Mock).mockResolvedValue({ type: '1' })
  expect(await getExcludePattern({ path: '/foo/bar' } as any)).toMatchSnapshot()
})
