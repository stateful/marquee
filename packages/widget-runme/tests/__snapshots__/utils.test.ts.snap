// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExcludePattern 1`] = `"**/dist/**,**/out/**,**/node_modules/**,**/.vscode-test/**,**/packages/**/build/**"`;

exports[`mapGitIgnoreToGlobFolders 1`] = `
[
  "**/dist/**",
  "**/out/**",
  "**/node_modules/**",
  "**/.vscode-test/**",
  "**/packages/**/build/**",
]
`;
